torchinductor/triton_ops/autotune.py

```diff
@@ -18,23 +23,99 @@
+    best_config = json.loads(open(cache_filename).read())
+    if best_config.get("configs_hash") != configs_hash:
+        return None
+
+    matching_configs = [
+        cfg
+        for cfg in configs
+        if all(val == best_config.get(key) for key, val in cfg.kwargs.items())
+    ]
+    if len(matching_configs) != 1:
+        return None
+
+    return matching_configs[0]
+
+
+def hash_configs(configs: List[Config]):
+    """
+    Hash used to check for changes in configurations
+    """
+    hasher = hashlib.sha256()
+    for cfg in configs:
+        hasher.update(
+            f"{sorted(cfg.kwargs.items())} {cfg.num_warps} {cfg.num_stages}\n".encode(
+                "utf-8"
+            )
+        )
+    return hasher.hexdigest()
 
 
-def autotune(configs, key, prune_configs_by=None, reset_to_zero=None):
+def autotune(
+    configs: List[Config],
+    key: List[str],
+    prune_configs_by=None,
+    reset_to_zero=None,
+    debug_label=None,
+    filename=None,
+):
     """
-    A copy of triton.autotune that calls our subclass above
+    A copy of triton.autotune that calls our subclass.  Our subclass
+    has additional debugging, error handling, and on-disk caching.
     """
+    configs = unique_configs(configs)
+    results = []
+
+    class Autotuner(triton.code_gen.Autotuner):
+        def _bench(self, *args, config, **kwargs):
+            try:
+                result = super()._bench(*args, config=config, **kwargs)
+                if debug_label:
+                    results.append(f"{config}: {result[0]}")
+                    if len(results) == len(configs):
+                        log.warning(
+                            "Autouning Results %s\n%s",
+                            debug_label,
+                            textwrap.indent("\n".join(results), " " * 4),
+                        )
+                return result
+            except triton.code_gen.OutOfResources as e:
+                log.warning("OutOfResources: %s %s", e, config)
+                return (float("inf"), float("inf"), float("inf"))
+
+    # on disk caching logic
+    if filename is not None and len(configs) > 1:
+        cache_filename = os.path.splitext(filename)[0] + ".best_config"
+        configs_hash = hash_configs(configs)
+
+        best_config = load_cached_autotuning(cache_filename, configs_hash, configs)
+        if best_config:
+            configs = [best_config]
+        else:
+
+            def call_and_store_cache(self, *args, **kwargs):
+                result = triton.code_gen.Autotuner.__call__(self, *args, **kwargs)
+                with open(cache_filename, "w") as fd:
+                    fd.write(
+                        json.dumps(
+                            {**self.best_config.kwargs, "configs_hash": configs_hash}
+                        )
+                    )
+                self.configs = [self.best_config]
+                del Autotuner.__call__  # uninstall this hook
```

Comment:
lol probably we should upstream it to triton to avoid tricks like this

Reply:
Yeah, agreed.  I'll take that as an action item.